util/fipstools/acvp/acvptool/subprocess/kdf-components.go

```diff
@@ -55,134 +65,282 @@ type kdfCompTestGroupResponse struct {
-			}
+	serverRandom, err := hex.DecodeString(test.ServerRandomHex)
+	if err != nil {
+		return err
+	}
 
-			const (
-				masterSecretLength = 48
-				masterSecretLabel  = "master secret"
-				extendedLabel      = "extended master secret"
-				keyBlockLabel      = "key expansion"
-			)
-
-			var outLenBytes [4]byte
-			binary.LittleEndian.PutUint32(outLenBytes[:], uint32(masterSecretLength))
-			var result [][]byte
-			switch k.algo {
-			case "kdf-components":
-				result, err = m.Transact(method, 1, outLenBytes[:], pms, []byte(masterSecretLabel), clientHelloRandom, serverHelloRandom)
-			case "TLS-v1.2":
-				result, err = m.Transact(method, 1, outLenBytes[:], pms, []byte(extendedLabel), sessionHash, nil)
-			default:
-				return nil, fmt.Errorf("unknown algorithm %q", k.algo)
-			}
-			if err != nil {
-				return nil, err
+	sessionHash, err := hex.DecodeString(test.SessionHashHex)
+	if err != nil {
+		return err
+	}
+
+	const (
+		masterSecretLength = 48
+		masterSecretLabel  = "master secret"
+		extendedLabel      = "extended master secret"
+		keyBlockLabel      = "key expansion"
+	)
+
+	var outLenBytes [4]byte
+	binary.LittleEndian.PutUint32(outLenBytes[:], uint32(masterSecretLength))
+	var result [][]byte
+	switch k.algo {
+	case "kdf-components":
+		result, err = m.Transact(method, 1, outLenBytes[:], pms, []byte(masterSecretLabel), clientHelloRandom, serverHelloRandom)
+	case "TLS-v1.2":
+		result, err = m.Transact(method, 1, outLenBytes[:], pms, []byte(extendedLabel), sessionHash, nil)
+	default:
+		return fmt.Errorf("unknown algorithm %q", k.algo)
+	}
+	if err != nil {
+		return err
+	}
+	binary.LittleEndian.PutUint32(outLenBytes[:], uint32(group.KeyBlockBits/8))
+	// TLS 1.0, 1.1, and 1.2 use a different order for the client and server
+	// randoms when computing the key block.
+	result2, err := m.Transact(method, 1, outLenBytes[:], result[0], []byte(keyBlockLabel), serverRandom, clientRandom)
+	if err != nil {
+		return err
+	}
+
+	response.Tests = append(response.Tests, kdfCompTestResponse{
+		ID:              test.ID,
+		MasterSecretHex: hex.EncodeToString(result[0]),
+		KeyBlockHex:     hex.EncodeToString(result2[0]),
+	})
+
+	return nil
+}
+
+func HandleSSH(test kdfCompTest, k *kdfComp, m Transactable, method string, group kdfCompTestGroup, response *kdfCompTestGroupResponse) error {
+	// See https://pages.nist.gov/ACVP/draft-celi-acvp-kdf-ssh.html
+	secretVal, err := hex.DecodeString(test.SecretValHex)
+	if err != nil {
+		return err
+	}
+
+	hashVal, err := hex.DecodeString(test.HashValHex)
+	if err != nil {
+		return err
+	}
+
+	sessionId, err := hex.DecodeString(test.SessionIdHex)
+	if err != nil {
+		return err
+	}
+
+	// NIST SP 800-135 says that the lengths of the IVs and encryption keys are determined by the supported block ciphers
+	// We can look at https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/json-files/kdf-components-ssh-1.0/internalProjection.json
+	// to determine what these lengths are
```

Comment:
If we can look at the data why do we need the hard coded maps?

Reply:
So the data for the hardcoded maps were found by looking through the NIST test vectors and determining the lengths of the IVs and keys for each cipher. For example if you look at the test vectors, you'll notice that the IV lengths for all vectors using AES-128 will be the same length, and the same will be the case for the encryption key and session ID for the other ciphers.

Getting that information is a matter of looking through the file, and since it never changes I don't think its worth finding a way to do that programmatically.

Reply:
`strlen(iv_buff) / 2` or something? Then you don't have to hard-code it for the cipher and it'll still  work if they add ChaCha20 or "AES II" in the future.